src/components/AllSpecies/AllSpecies.tsx

```diff
@@ -8,16 +8,64 @@ type AllSpeciesProps = {
   characters: ModifiedCharacter[]
 }
 
+type CharacterState = {
+  originalCharacters: ModifiedCharacter[]
+  modifiedCharacters: ModifiedCharacter[]
+  selectedValue: string
```

Comment:
I think you can narrow this type to a union of string literals. In this use case, you know the selected value is always either `height` or `mass`, so I'd recommend you create an enumerator-like object so that you can couple all the logic looking at the single source of truth. This is phenomenally effective when all the logic is based on a single factor/domain. In this case, we know they all are trying to "sort" characters with these keys, so it's ok to couple them together to look at the single place. 

This is just a pseudo code so don't count on it lol but I hope you get the idea.
```typescript

// enum like object - make sure to leave a comment about why you use this object as it has multiple dependencies. Means that a single change on this object can cause a big mess 
const SORT_FACTOR = {
  MASS: "mass",
  HEIGHT: "height",
  DEFAULT: "default" // you might not need this value
} as cosnt

// "mass" | "height" | "default"
type SortFactor = typeof SORT_FACTOR[keyof typeof SORT_FACTOR]

type CharacterState = {
  originalCharacters: ModifiedCharacter[]
  modifiedCharacters: ModifiedCharacter[]
  selectedValue: SortFactor // here, you narrowed the type from "string" to "mass" | "height" | "default"
}

// this function now can accept "mass" | "height" | "default". You might want to handle the case "default" explicitly
const sort = (characters: ModifiedCharacter[], key: SortFactor) =>
  characters.sort((a, b) => (b[key] || 0) - (a[key] || 0))

[...]

{
    Object.keys(SORT_FACTOR).map(sortFactorKey => (
        // You might need some logic or CSS trick to change the value to the title case
        <option value={SORT_FACTOR[sortFactorKey]}>{SORT_FACTOR[sortFactorKey]}</option>
     ))
}
```

Reply:
@koh1project Pinging you just in case as this is quite a good example of TS-specific abstraction logic👀

Reply:
Thank you @kei95 for mentionning me. I remember that I used the `typeof SORT_FACTOR[keyof typeof SORT_FACTOR]` usage. I used it for reducers, and also abstracted the logic as `type ValueOf<T> = T[keyof T];`

@Lada496 
I quickly wrote the characterReducer with ValueOf usage in [playground](https://www.typescriptlang.org/play?ssl=1&ssc=30&pln=1&pc=1#code/C4TwDgpgBAaghgGwK4QPIDMA8AVAfFAXimwG0BrCEAe3WIF0BuAKBYGMqA7AZ2CgGEAFnABOcVsAjCAguICWnbOAhdCUAN5MoUAGIBJADLYAogCUAXFADk6WQgnDLAGk1QpAOVTYAEqYuW4HFTAApKWTAC+UHAq7NzAzCygkPxCouKSMsDyHKoaWkkQFvDIaFgFNCkiYvaZ2YqQXLjMWhAAbhAcwBYAylTCwNrVfc1QYHAgCFRwACYA-BYBIMzhCQD0q1AAmlRIUKwBUQhcVFBIXNCyvMAnvf2D4n1M61AFULcDQ8KqxSgYmOW0bqoEzYAD62ikfGwwNwbE4PD2qWqkhMEGmSFYklUAAoeHAJBZBFV0sJusB8RBHFE5JxCUiSbVOABKOnE+xkimEfB5LRQLgAd0urAEUGx1WyADoCkz1C4tPtzpU0jUaRx6soJXpDL45bytMIIMAkMIcjy9Xq+rIAOayDiIInKyRcCwkNThOhU554iQSy02u0IB3I4RcZzm80AWyo01kNjRQZJzqgrvdno2HAg-IAstHY7J4-T7FxdeHzggIOlpj9CtSspwJW0OsAw+GoOFdQroAmVXW1UouBL3J4fOYS7yDUaTXzyRIRlpphB0HAkHYzGP9YbjTlvRA57z2ysWCxntgQlAAEYQSb8vkV3tQWQqdDCKgRqAAaSMuksMVfEabEqEAQwEgaBYHgRBoFMLECJAiC4KQtCJi5C4WZSN03QWAARBG0RcFhLY+LoADiXjYNhITWgIwAES4AAiRgQgAqoY2ELkuK40VAzzULsEZUbwgSCRAaIvAIj5QK0iAoBEUS-nEawbDheFYVAAA+UBYZRVrUapGlYexy52FhTCvO89zXF8RAAm8wJghCULAuQlAVDZcH2YhwJ0IkShKsGHISChWh+ra9qFk6Lput5WhRjGcbTN2EXJlFLhlneaLVj0fQfA8XzPCEBpUrxUB2sIL78qJwTQK8z6vppPDCLaVqqdcmm4Vw+HqZp2m6V1BmLkZNEREwQA), as if I applied it.

Reply:
> Good change, I left a bunch of nitpicky comments but overall it looks good. Feel free to skip any comments if you feel any of them are unnecessary! I found that this page has two scrollable views but it's a bit unpredictable as characters' view can be scrolled only within its width, so when I scroll the further right side of this page it scrolls categories where I(a user) expect the nested view to be scrolled.
> 
> ![2023-01-08 11 04 45](https://user-images.githubusercontent.com/44686790/211215513-3703ebe4-be40-4d3b-a5b9-93750bfa6e96.gif) [ ![2023-01-08 11 04 45](https://user-images.githubusercontent.com/44686790/211215513-3703ebe4-be40-4d3b-a5b9-93750bfa6e96.gif) ](https://user-images.githubusercontent.com/44686790/211215513-3703ebe4-be40-4d3b-a5b9-93750bfa6e96.gif) [ ](https://user-images.githubusercontent.com/44686790/211215513-3703ebe4-be40-4d3b-a5b9-93750bfa6e96.gif)

Added an issue for this
https://github.com/Lada496/star-wars-list/issues/24